components/script/script_runtime.rs

```diff
@@ -129,6 +140,118 @@ unsafe extern "C" fn enqueue_job(cx: *mut JSContext,
     }), false)
 }
 
+#[allow(unsafe_code)]
+/// https://html.spec.whatwg.org/multipage/#the-hostpromiserejectiontracker-implementation
+unsafe extern "C" fn promise_rejection_tracker(
+    cx: *mut JSContext,
+    promise: HandleObject,
+    state: PromiseRejectionHandlingState,
+    _data: *mut c_void
+) {
+    // TODO: Step 2 - If script's muted errors is true, terminate these steps.
+
+    // Step 3.
+    let global = GlobalScope::from_context(cx);
+
+    wrap_panic(AssertUnwindSafe(|| {
+        match state {
+            // Step 4.
+            PromiseRejectionHandlingState::Unhandled => {
+                global.add_uncaught_rejection(promise);
+            },
+            // Step 5.
+            PromiseRejectionHandlingState::Handled => {
+                // Step 5-1.
+                if global.get_uncaught_rejections().borrow().contains(&Heap::boxed(promise.get())) {
+                    global.remove_uncaught_rejection(promise);
+                    return;
+                }
+
+                // Step 5-2.
+                if !global.get_consumed_rejections().borrow().contains(&Heap::boxed(promise.get())) {
+                    global.add_consumed_rejection(promise);
+                    return;
+                }
+
+                // Step 5-3.
+                global.remove_consumed_rejection(promise);
+
+                // TODO: Step 5-4 - Queue a task to fire `rejectionhandled` event
+            }
+        };
+    }), ());
+}
+
+#[allow(unsafe_code, unrooted_must_root)]
+/// https://html.spec.whatwg.org/multipage/#notify-about-rejected-promises
+pub fn notify_about_rejected_promises(global: &GlobalScope) {
+    unsafe {
+        let cx = global.get_cx();
+
+        // Step 2.
+        if global.get_uncaught_rejections().borrow().len() > 0 {
+            // Step 1.
+            let uncaught_rejections: Vec<TrustedPromise> = global.get_uncaught_rejections()
+                .borrow()
+                .iter()
+                .map(|promise| {
+                    let promise = Promise::new_with_js_promise(Handle::from_raw(promise.handle()), cx);
+
+                    TrustedPromise::new(promise)
+                })
+                .collect();
+
+            // Step 3.
+            global.get_uncaught_rejections().borrow_mut().clear();
+
+            let target = Trusted::new(global.upcast::<EventTarget>());
+
+            // Step 4.
+            global.dom_manipulation_task_source().queue(
+                task!(unhandled_rejection_event: move || {
+                    let target = target.root();
+                    let cx = target.global().get_cx();
+
+                    for promise in uncaught_rejections {
+                        // TODO: Step 4-1 - if [[GetPromiseIsHandled]] is true, return
```

Comment:
@asajeffrey cc @jdm 

As mentioned in the spec, we'll need to check the internal slot `[[GetPromiseIsHandled]]` for the promise.

But I found we don't have `GetPromiseIsHandled` for getting the `PromiseIsHandled` slot like `GetPromiseResult` which is used to get the fulfilled `value` or rejected `reason`, how can I contribute to `mozjs` to add the function? or, could you help me about it? Thanks!

Reply:
Want to add a GetPromiseIsHandled API? You should be able to use much of the same implementation as GetPromiseResult.

Reply:
Yes, let me try to check how to make a `GetPromiseIsHandled` API and I will send a PR to `rust-mozjs`! cc @asajeffrey